staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go

```diff
@@ -0,0 +1,51 @@
+package request
+
+import (
+	"context"
+)
+
+type shardKey int
+
+const (
+	// shardKey is the context key for the request.
+	shardContextKey shardKey = iota
+)
+
+// Shard holds data that describes a shard.
+type Shard struct {
+	// Name is the name of the shard.
+	Name string
+
+	// If true the query applies to all shards
+	Wildcard bool
+}
+
+// Empty returns true if the name of the shard is empty.
+func (s *Shard) Empty() bool {
+	return s.Name == ""
+}
+
+// WithShard returns a context that holds the given shard.
+func WithShard(parent context.Context, shard Shard) context.Context {
+	return context.WithValue(parent, shardContextKey, shard)
+}
+
+// ShardFrom returns the value of the shard key in the context, or nil if there is no shard key.
+func ShardFrom(ctx context.Context) *Shard {
+	shard, ok := ctx.Value(shardContextKey).(Shard)
+	if !ok {
+		return nil
+	}
+	return &shard
+}
+
+// ValidShardFrom returns the value of the shard key from the context.
+// An empty value is returned if there's no shard key,
+// or if the shard name is empty, and it wasn't a wildcard request.
+func ValidShardFrom(ctx context.Context) (*Shard, error) {
```

Comment:
why "Valid"? ShardFromOrEmpty would be more idiomatic

Reply:
true